### ISSUE TITLE
Why VTCM is slower than others(Heap/Stack/Register)(target=hexagon-32-qurt-hvx_128)

### ISSUE THREAD

**Issue by gavin838** [halide/Halide#6187]
When I using 'store_in(MemoryType::VTCM)' in schedule, I get larger cycle. I think VTCM is the fastest buffer, did I get it wrong?
My test cycle(8350):
store_in(MemoryType::VTCM): **10589** microseconds
store_in(MemoryType::Heap): **150** microseconds
store_in(MemoryType::Stack): **40** microseconds
store_in(MemoryType::Register): **40** microseconds
store_in(MemoryType::GPUShared): **40** microseconds

**Comment by dsharletg** [comment]
@aankit-ca @pranavb-ca should chime in on this one. But in general I'd expect this to depend on what you are doing. Based on the performance data you shared, my guess is that allocation overhead is dominating your benchmark (otherwise Heap and Stack should probably perform similarly).

Even after fixing that though, I'd expect storing things in registers to be best, assuming register pressure/spills aren't a problem.

**Comment by aankit-ca** [comment]
We have noticed this in the past, allocating memory on VTCM is much slower. I guess the expectation is that we allocate the memory once and reuse it wisely. But I can try to see if we have some numbers to compare HAP_request_VTCM and malloc.

**Comment by dsharletg** [comment]
Yes, I think it's reasonable that there is some cost to allocation, it just seems like it might be excessive compared to heap allocation (it seems like that should be at least the same order of magnitude).

**Comment by gavin838** [comment]
@dsharletg My algorithm is gaussian blur(3x3), it shound be compute-bound op.
Here is my algorithm and schedule:
`void generate() {
        Func repeat_3{"repeat_3"};
        repeat_3 = BoundaryConditions::repeat_edge(input);
        Func rows{"rows"}, rows_t{"rows_t"}, cols{"cols"};
        rows(x, y) = repeat_3(__X_3,     y) + 2* repeat_3(__X_3 + 1, y) + repeat_3(__X_3 + 2,     y);
        rows_t(x, y) = rows(x, __Y_3) + rows(x, __Y_3+2);
        output(x, y) = (8 + rows_t(x, y) + 2 * rows(x, __Y_3+1)) >> 4;
    }
inline void apply_schedule_test_halide(
    ::Halide::Pipeline pipeline,
    ::Halide::Target target
) {
    using ::Halide::Func;
    using ::Halide::MemoryType;
    using ::Halide::RVar;
    using ::Halide::TailStrategy;
    using ::Halide::Var;
    Func output = pipeline.get_func(5);
    Func rows_t = pipeline.get_func(4);
    Func rows = pipeline.get_func(3);
    Func repeat_edge = pipeline.get_func(2);
    Func lambda_0 = pipeline.get_func(1);
    Func input_im = pipeline.get_func(0);
    Var _0(repeat_edge.get_schedule().dims()[0].var);
    Var _0i("_0i");
    Var _1(repeat_edge.get_schedule().dims()[1].var);
    Var x(output.get_schedule().dims()[0].var);
    Var xi("xi");
    Var xii("xii");
    Var y(output.get_schedule().dims()[1].var);
    Var yi("yi");
    output
        .split(x, x, xi, 128, TailStrategy::ShiftInwards)
        .split(y, y, yi, 8, TailStrategy::ShiftInwards)
        .split(xi, xi, xii, 64, TailStrategy::ShiftInwards)
        .unroll(yi)
        .vectorize(xii)
        .align_storage(x, 64)
        .compute_root()
        .reorder({xii, yi, xi, x, y})
        .fuse(x, y, x)
        .parallel(x)
        .prefetch(input_im, yi, 2);
    rows_t
        .store_in(MemoryType::Register)
        .split(x, x, xi, 32, TailStrategy::RoundUp)
        .unroll(x)
        .vectorize(xi)
        .align_storage(x, 32)
        .compute_at(output, yi)
        .reorder({xi, x, y});
    rows
        .store_in(MemoryType::Register)
        .split(x, x, xi, 64, TailStrategy::RoundUp)
        .unroll(y)
        .vectorize(xi)
        .align_storage(x, 64)
        .compute_at(output, yi)
        .store_at(output, xi)
        .reorder({xi, x, y});
    repeat_edge
        .store_in(MemoryType::Register)
        .split(_0, _0, _0i, 64, TailStrategy::ShiftInwards)
        .unroll(_0)
        .unroll(_1)
        .vectorize(_0i)
        .align_storage(_0, 64)
        .compute_at(output, yi)
        .store_at(output, xi)
        .reorder({_0i, _0, _1});

}`

**Comment by dsharletg** [comment]
It looks like you could change the `store_at(output, xi)` to `store_at(output, x)`, which might better amortize the cost of allocation.

However, in this case, if storing in registers works, that's probably going to be the best option.

**Comment by gavin838** [comment]
Thanks, @dsharletg , do you know the bandwith of buffer(L2, VTCM, Heap, Stack, Register...) on Qualcomm8350 ? Just the order of them, like Register > Heap > Stack > VTCM > L2, I can't find the order accoriding Qualcomm infomation.

**Comment by dsharletg** [comment]
I would think the order is Register > VTCM >= L2 > Stack = Heap
